### ISSUE TITLE
Handle both encodings of optional ASN.1 NULL fields

### ISSUE THREAD

**Issue by DemiMarie-parity** [briansmith/webpki#123]
There are many places where an AlgorithmIdentifier can be specified either without parameters or with NULL parameters.  See [RFC 8017](https://tools.ietf.org/html/rfc8107).

**Comment by briansmith** [comment]
> See [RFC 8017](https://tools.ietf.org/html/rfc8107).

I think you are referring to another RFC.

> This only affects the RSASSA-PSS-* algorithms, specifically src/data/alg-rsa-pss-sha{256,384,512}.der.

What do you think is the correct encoding for those files? Could you share (perhaps in a PR) replacements for those files that are encoded correctly?

One of the goals of the webpki project is to be as strict about parsing as practical. As part of this, we try to avoid accepting two different ways of encoding something; instead we try to accept only "the one true way." I think there are some exceptions to this, for compatibility reasons; I think some of those compatibility workarounds can now actually be removed now that browser CA programs are enforcing that valid encodings being used for new certificates. Thus, I'm not eager to accept two different ways of encoding anything. However, this isn't absolute; we might add such logic if/when we feel like we have a good reason to do so.

**Comment by briansmith** [comment]
See previous discussion https://github.com/briansmith/webpki/issues/25#issuecomment-261113319.

**Comment by briansmith** [comment]
See https://github.com/mozilla/pkipolicy/blob/2.7/rootstore/policy.md#511-rsa and

**Comment by DemiMarie-parity** [comment]
> > See [RFC 8017](https://tools.ietf.org/html/rfc8107).
> 
> I think you are referring to another RFC.

The link was wrong, sorry.  Fixed.
 
> > This only affects the RSASSA-PSS-* algorithms, specifically src/data/alg-rsa-pss-sha{256,384,512}.der.
> 
> What do you think is the correct encoding for those files? Could you share (perhaps in a PR) replacements for those files that are encoded correctly?

The files are encoded correctly, but there are three other encodings for each of them which I also believe are valid.

> One of the goals of the webpki project is to be as strict about parsing as practical. As part of this, we try to avoid accepting two different ways of encoding something; instead we try to accept only "the one true way." I think there are some exceptions to this, for compatibility reasons; I think some of those compatibility workarounds can now actually be removed now that browser CA programs are enforcing that valid encodings being used for new certificates. Thus, I'm not eager to accept two different ways of encoding anything. However, this isn't absolute; we might add such logic if/when we feel like we have a good reason to do so.

In this case, it appears that the ambiguity is required by a MUST in RFC8017, which states that the parameters for the SHA{256,384,512} AlgorithmIds may be NULL or absent, and requires that both be accepted.

**Comment by briansmith** [comment]
Have you seen https://tools.ietf.org/html/rfc5756 says:

> When the id-RSASSA-PSS object identifier appears in the
   TBSCertificate or TBSCertList signature algorithm field, then the
   RSASSA-PSS-params structure MUST be included in the TBSCertificate or
   TBSCertList signature parameters field.

...

>   All certificates and CRLs signed by a CA that supports the id-RSASSA-
   PSS algorithm MUST include the RSASSA-PSS-params in the
   signatureAlgorithm parameters in Certificate and CertList structures,
   respectively.

Here's my position: If a CA is going to issue certificates signed using PSS, then it needs to conform to RFC5756 and we should reject any certificates that do not conform to RFC5756. And, I think this is what webpki is implementing here.

Am I misunderstanding something?

What software is generating certificates that do not conform to RFC5756?

**Comment by briansmith** [comment]
See https://github.com/mozilla/pkipolicy/pull/183. This is what webpki aims to enforce. If webpki is enforcing something different, then that is a bug that should be fixed. If it is enforcing that the text requires, then webpki is working as designed, and we'd need a compelling reason to change it.

**Comment by DemiMarie-parity** [comment]
> Am I misunderstanding something?

Yes.  From what I can tell, RSASSA-PSS-params can be encoded in four different ways: it has two instances of hash function `AlgorithmId`s, and each of those can be encoded with or without the optional NULL. Furthermore, the encoding *without* the NULL appears to be preferred, and is indeed what OpenSSL generates.  Go (incorrectly, IMO) includes the NULL, but both encodings MUST be accepted.

> What software is generating certificates that do not conform to RFC5756?

None that I am aware of.

**Comment by briansmith** [comment]
> . Furthermore, the encoding without the NULL appears to be preferred

OK, I didn't realize that. Could you please site a spec that says the encoding without the NULL is preferred? I found https://tools.ietf.org/html/rfc4055#section-6 which says the NULL is required for both digest algorithm identifiers and https://github.com/mozilla/pkipolicy/blob/master/rootstore/policy.md#511-rsa which says the NULL is required for both.

> and is indeed what OpenSSL generates. 

Please provide working commands for generating such a certificate from OpenSSL.

My inclination is to file a bug report against OpenSSL to make it work like Go and become compatible with webpki.

**Comment by briansmith** [comment]
This is blocked until we get the OpenSSL commands to reproduce.

**Comment by DemiMarie-parity** [comment]
> > . Furthermore, the encoding without the NULL appears to be preferred
> 
> OK, I didn't realize that. Could you please site a spec that says the encoding without the NULL is preferred? I found https://tools.ietf.org/html/rfc4055#section-6 which says the NULL is required for both digest algorithm identifiers and https://github.com/mozilla/pkipolicy/blob/master/rootstore/policy.md#511-rsa which says the NULL is required for both.

I am not an expert on X.509, but my source is https://tools.ietf.org/html/rfc8017#appendix-C which states:

```
-- When id-sha1, id-sha224, id-sha256, id-sha384, id-sha512,
-- id-sha512-224, and id-sha512-256 are used in an
-- AlgorithmIdentifier, the parameters (which are optional) SHOULD be
-- omitted, but if present, they SHALL have a value of type NULL.
-- However, implementations MUST accept AlgorithmIdentifier values
-- both without parameters and with NULL parameters.

-- Exception: When formatting the DigestInfoValue in EMSA-PKCS1-v1_5
-- (see Section 9.2), the parameters field associated with id-sha1,
-- id-sha224, id-sha256, id-sha384, id-sha512, id-sha512-224, and
-- id-sha512-256 SHALL have a value of type NULL.  This is to
-- maintain compatibility with existing implementations and with the
-- numeric information values already published for EMSA-PKCS1-v1_5,
-- which are also reflected in IEEE 1363a.
```

While I could be misreading the ASN.1 module, this text seems clear.

This script can reproduce the OpenSSL behavior:

    #!/bin/sh --
    set -euf
    export LC_ALL=en_US.UTF-8
    case $0 in
       (/*) dir=${0%/*}/;;
       (*/*) dir=./${0%/*};;
       (*) dir=.;;
    esac
    cd -- "$dir"
    tmpdir=$(mktemp -d)
    trap '
    set +fe
    shred -- "$tmpdir"/*.key
    rm -rf -- "$tmpdir"' EXIT
    conf=$tmpdir/openssl.cnf ecparams=$tmpdir/params.txt
    cakey=ca.key testkey=$tmpdir/testing.key csr=$tmpdir/testing.csr
    cat > "$conf" <<'EOF'
    [ req ]
    x509_extensions = v3_ca
    distinguished_name = req_distinguished_name
    default_md = sha256
    encrypt_key = no
    prompt = no
    string_mask = utf8only
    utf8 = yes
    
    [ req_distinguished_name ]
    CN = dummy
    
    [ v3_ca ]
    subjectKeyIdentifier = hash
    basicConstraints = critical,CA:TRUE,pathlen:0
    keyUsage = critical,cRLSign,keyCertSign
    # webpki does’t understand this
    authorityInfoAccess = critical,OCSP;URI:https://example.invalid
    
    [ usr_cert ]
    basicConstraints = critical,CA:FALSE
    subjectKeyIdentifier = hash
    authorityKeyIdentifier = keyid:always
    keyUsage = critical,nonRepudiation,digitalSignature
    extendedKeyUsage = critical,serverAuth,clientAuth
    # Do not use a subject alt name, as libp2p doesn’t either.
    # subjectAltName = DNS:localhost
    # webpki does’t understand this
    authorityInfoAccess = critical,OCSP;URI:https://example.invalid
    EOF
    
    openssl ecparam -name prime256v1 > "$ecparams"
    
    openssl req \
       -x509 \
       -newkey rsa \
       -batch \
       -days 3000 \
       -outform der \
       -subj '/CN=dummy_ca' \
       -multivalue-rdn \
       -out ca.crt \
       -keyout "$cakey" \
       -config "$conf"
    
    openssl req \
       -newkey "ec:$ecparams" \
       -batch \
       -out "$csr" \
       -keyout "$testkey" \
       -keyform der \
       -config "$conf"
    
    openssl x509 \
       -req \
       -sha256 \
       -outform der \
       -inform der \
       -CAkey "$cakey" \
       -CAform der \
       -days 3000 \
       -CA ca.crt \
       -CAcreateserial \
       -CAserial "$tmpdir/ca.srl" \
       -clrext \
       -extfile "$conf" \
       -extensions usr_cert \
       -in "$csr" \
       -out testing.crt

**Comment by briansmith** [comment]
I am more open to doing this. The Chromium certificate verifier, the mozilla::pkix, and BoringSSL have all been changed to have this flexibility, and webpki probably can't afford to be stricter than them. I think we should consider changing the Mozilla PKI policy to allowing the other form too. I seem to remember a recent RFC attempts to clarify the correct encoding and that RFC seems to have chosen the other value. I think when there is such a dispute then there should be a preference for choosing the shortest encoding.

**Comment by DemiMarie-parity** [comment]
I agree.  Using a shorter encoding saves bytes on the wire and has no security disadvantages.
